federation/pkg/federation-controller/replicaset/replicasetcontroller.go

```diff
@@ -0,0 +1,454 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replicaset
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	fed "k8s.io/kubernetes/federation/apis/federation"
+	fedv1 "k8s.io/kubernetes/federation/apis/federation/v1beta1"
+	fedclientset "k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_4"
+	//kubeclientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_4"
+	planner "k8s.io/kubernetes/federation/pkg/federation-controller/replicaset/planner"
+	fedutil "k8s.io/kubernetes/federation/pkg/federation-controller/util"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/meta"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	extensionsv1 "k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	// schedule result was put into annotation in a format of "clusterName:replicas[/clusterName:replicas]..."
+	ExpectedReplicasAnnotation = "kubernetes.io/expected-replicas"
+	allClustersKey             = "THE_ALL_CLUSTER_KEY"
+	UserAgentName              = "Federation-replicaset-Controller"
+)
+
+var (
+	replicaSetReviewDelay   = 10 * time.Second
+	clusterAvailableDelay   = 20 * time.Second
+	clusterUnavailableDelay = 60 * time.Second
+)
+
+func encodeScheduleResult(scheduleResult map[string]int64) string {
```

Comment:
0 it's not possible since clusters may go away and that part of information went missing.

1 the current planner is stable (given the same parameter), however clusters may come and go and the results may different which will cause re-balancing (for all the federation replicasets).
besides, enforce a scheduler to be stable (in a dynamic environment) removes the possibility of more sophisticated schedulers from for example take the cluster load into account or do parallel scheduling, etc.
1.1 actually we don't want the re-schedule to be the same as a cluster may fail fulfilling the required replicas and we should use that information for the next schedule.

2 again, I cannot count on a re-schedule to be the exact output, now or later. an re-schedule should only happen when needed.

3 yes these is the cache, but it still need to iterate all the clusters (even we could know there were no replicasets interested), and I cannot retrieve this info when the cluster is gone

4 see point 1